Monika After Story/game/zz_submods.rpy

```diff
@@ -1,102 +1,472 @@
 init -999:
     default persistent._mas_submod_version_data = dict()
 
-init 10 python:
+init 10 python in mas_submod_utils:
     #Run updates if need be
-    store.mas_submod_utils.Submod._checkUpdates()
-
-init -989 python:
-    #Log initialized submods
-    if store.mas_submod_utils.submod_map:
-        mas_submod_utils.submod_log.info(
-            "\nINSTALLED SUBMODS:\n{0}".format(
-                ",\n".join(
-                    ["    '{0}' v{1}".format(submod.name, submod.version) for submod in store.mas_submod_utils.submod_map.values()]
-                )
-            )
-        )
+    _Submod._checkUpdates()
 
-    #Run dependency checks
-    store.mas_submod_utils.Submod._checkDependencies()
+init -999 python in mas_submod_utils:
+    # Init submods
+    _load_submods()
 
-init -991 python in mas_submod_utils:
+init -1000 python in mas_submod_utils:
+    import glob
     import re
+    import os
+    import json
+    import sys
+    from urllib.parse import urlparse
+    from typing import Literal, Optional
+    from collections.abc import Iterator
+
+    import pydantic
+
     import store
-    # import sys
-    # import traceback
+    from store import (
+        config,
+        persistent,
+        mas_utils,
+        mas_logging,
+        _mas_loader
+    )
+
+
+    submod_log = mas_logging.init_log("submod_log")
+
+    # NOTE: ALWAYS UPDATE VERSION IF YOU CHANGE HEADER FORMAT
+    HEADER_VERSION = 1
+
+    HEADER_GLOB = "**/header.json"
+    SUBMODS_DIR = "Submods"
+
+    SANE_STR_PATTERN = re.compile(r'^[a-zA-Z_\u00a0-\ufffd][ 0-9a-zA-Z_\u00a0-\ufffd]*$')
```

Comment:
Given the form of validation we use, this is an allow list.

Out of curiosity, why do we allow zero-width spaces and replacement characters? Given the fields that are validated by this pattern will be merged together to create the submod's update label, this will cause an invalid label to be generated, making it impossible for any submod using a name with these chars to be able to support version update scripts.

Unless there's a reason these are here, we should avoid supporting these chars.

Reply:
Maybe they could be stripped when making a label out of them?